File: visualization/core/model-config.js
/**
 * Model configuration loader.
 * Loads model architecture and settings from config/models/*.yaml
 *
 * Usage:
 *     await modelConfig.load('gemma-2-2b-it');
 *
 *     modelConfig.get('num_hidden_layers');  // 26
 *     modelConfig.get('sae.available');      // true
 *     modelConfig.getNumLayers();            // 26
 *     modelConfig.getSaePath(16);            // 'sae/gemma-scope.../layer_16_...'
 */

class ModelConfig {
    constructor() {
        this.config = null;
        this.modelId = null;
        this.loaded = false;
        this._cache = {};
    }

    /**
     * Load model config from YAML file.
     * @param {string} modelId - Model identifier (e.g., 'gemma-2-2b-it')
     */
    async load(modelId) {
        // Normalize: google/gemma-2-2b-it -> gemma-2-2b-it
        if (modelId.includes('/')) {
            modelId = modelId.split('/').pop().toLowerCase();
        }

        // Return cached if same model
        if (this.modelId === modelId && this.loaded) {
            return this.config;
        }

        // Check cache
        if (this._cache[modelId]) {
            this.config = this._cache[modelId];
            this.modelId = modelId;
            this.loaded = true;
            return this.config;
        }

        const response = await fetch(`/config/models/${modelId}.yaml`);
        if (!response.ok) {
            throw new Error(`Failed to load model config for '${modelId}': ${response.status}`);
        }

        const yamlText = await response.text();
        this.config = jsyaml.load(yamlText);
        this.modelId = modelId;
        this.loaded = true;
        this._cache[modelId] = this.config;

        return this.config;
    }

    /**
     * Load model config for an experiment.
     * Reads experiment's config.json to determine model, then loads model config.
     * @param {string} experiment - Experiment name
     */
    async loadForExperiment(experiment) {
        let modelId = experiment;

        // Try to read experiment config
        try {
            const response = await fetch(`/experiments/${experiment}/config.json`);
            if (response.ok) {
                const expConfig = await response.json();
                if (expConfig.model) {
                    modelId = expConfig.model;
                }
            }
        } catch (e) {
            // Fall back to experiment name as model ID
            console.log(`No experiment config for ${experiment}, using experiment name as model ID`);
        }

        return this.load(modelId);
    }

    /**
     * Get a config value by dot-separated key.
     * @param {string} key - Dot-separated key like 'sae.available'
     * @returns {*} Config value
     */
    get(key) {
        if (!this.loaded) {
            throw new Error('ModelConfig not loaded. Call await modelConfig.load() first.');
        }

        let value = this.config;
        for (const k of key.split('.')) {
            if (value === undefined || value === null) return undefined;
            value = value[k];
        }
        return value;
    }

    /**
     * Get the full config object.
     * @returns {Object}
     */
    getAll() {
        if (!this.loaded) {
            throw new Error('ModelConfig not loaded. Call await modelConfig.load() first.');
        }
        return this.config;
    }

    // Convenience accessors

    getNumLayers() {
        return this.get('num_hidden_layers');
    }

    getHiddenSize() {
        return this.get('hidden_size');
    }

    getNumHeads() {
        return this.get('num_attention_heads');
    }

    getNumKvHeads() {
        return this.get('num_key_value_heads');
    }

    getResidualSublayers() {
        return this.get('residual_sublayers') || ['input', 'after_attn', 'output'];
    }

    getNumSublayers() {
        return this.getResidualSublayers().length;
    }

    getDefaultMonitoringLayer() {
        return this.get('defaults.monitoring_layer') || Math.floor(this.getNumLayers() / 2);
    }

    isSaeAvailable() {
        return this.get('sae.available') || false;
    }

    getSaeDownloadedLayers() {
        return this.get('sae.downloaded_layers') || [];
    }

    /**
     * Get SAE path for a specific layer.
     * @param {number} layer - Layer index
     * @returns {string|null} Path to SAE directory, or null if not available
     */
    getSaePath(layer) {
        if (!this.isSaeAvailable()) return null;

        const downloadedLayers = this.getSaeDownloadedLayers();
        if (!downloadedLayers.includes(layer)) return null;

        const basePath = this.get('sae.base_path');
        const template = this.get('sae.layer_template');
        const layerDir = template.replace('{layer}', layer);

        return `${basePath}/${layerDir}`;
    }
}

// Global instance
const modelConfig = new ModelConfig();
window.modelConfig = modelConfig;

File: utils/model_config.py
"""
Model configuration loader.
Loads model architecture and settings from config/models/*.yaml

Input: Model ID (e.g., 'gemma-2-2b-it') or experiment name
Output: Dict with model config (architecture, SAE info, defaults)

Usage:
    from utils.model_config import get_model_config, get_config_for_experiment

    # Direct model lookup
    config = get_model_config('gemma-2-2b-it')
    config['num_hidden_layers']  # 26
    config['sae']['available']   # True

    # From experiment (reads experiment's config.json to get model)
    config = get_config_for_experiment('gemma-2-2b-it')
"""

import json
import yaml
from pathlib import Path
from typing import Dict, Optional, Any

_cache: Dict[str, dict] = {}
_models_dir = Path(__file__).parent.parent / "config" / "models"
_experiments_dir = Path(__file__).parent.parent / "experiments"


def get_model_config(model_id: str) -> dict:
    """
    Load model config by ID.

    Args:
        model_id: Model identifier matching YAML filename (e.g., 'gemma-2-2b-it')

    Returns:
        Dict with model configuration

    Raises:
        FileNotFoundError: If model config doesn't exist
    """
    if model_id in _cache:
        return _cache[model_id]

    # Try exact match first
    config_path = _models_dir / f"{model_id}.yaml"

    if not config_path.exists():
        # Try normalizing: google/gemma-2-2b-it -> gemma-2-2b-it
        if '/' in model_id:
            model_id = model_id.split('/')[-1].lower()
            config_path = _models_dir / f"{model_id}.yaml"

    if not config_path.exists():
        raise FileNotFoundError(f"No model config found for '{model_id}' at {config_path}")

    with open(config_path) as f:
        config = yaml.safe_load(f)

    _cache[model_id] = config
    return config


def get_config_for_experiment(experiment: str) -> dict:
    """
    Load model config for an experiment.
    Reads experiment's config.json to determine which model, then loads model config.

    Args:
        experiment: Experiment name (e.g., 'gemma-2-2b-it')

    Returns:
        Dict with model configuration
    """
    exp_config_path = _experiments_dir / experiment / "config.json"

    if exp_config_path.exists():
        with open(exp_config_path) as f:
            exp_config = json.load(f)
        model_id = exp_config.get('model', experiment)
    else:
        # Fall back to experiment name as model ID
        model_id = experiment

    # Normalize HF ID to our config name
    # google/gemma-2-2b-it -> gemma-2-2b-it
    if '/' in model_id:
        model_id = model_id.split('/')[-1].lower()

    return get_model_config(model_id)


def list_available_models() -> list:
    """List all available model configs."""
    return [p.stem for p in _models_dir.glob("*.yaml")]


# Convenience accessors
def get_num_layers(model_or_experiment: str) -> int:
    """Get number of hidden layers for a model/experiment."""
    try:
        config = get_config_for_experiment(model_or_experiment)
    except FileNotFoundError:
        config = get_model_config(model_or_experiment)
    return config['num_hidden_layers']


def get_hidden_size(model_or_experiment: str) -> int:
    """Get hidden dimension for a model/experiment."""
    try:
        config = get_config_for_experiment(model_or_experiment)
    except FileNotFoundError:
        config = get_model_config(model_or_experiment)
    return config['hidden_size']


def get_sae_path(model_or_experiment: str, layer: int) -> Optional[Path]:
    """
    Get SAE path for a specific layer, if available.

    Returns:
        Path to SAE directory, or None if SAE not available for this model/layer
    """
    try:
        config = get_config_for_experiment(model_or_experiment)
    except FileNotFoundError:
        config = get_model_config(model_or_experiment)

    sae = config.get('sae', {})
    if not sae.get('available', False):
        return None

    downloaded = sae.get('downloaded_layers', [])
    if layer not in downloaded:
        return None

    base_path = Path(__file__).parent.parent / sae['base_path']
    layer_dir = sae['layer_template'].format(layer=layer)
    return base_path / layer_dir
